contracts/competition/Competition.sol

```diff
@@ -0,0 +1,169 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.6.0;
+
+import "../referrals/interfaces/IReferralStorage.sol";
+import "../access/Governable.sol";
+
+contract Competition is Governable
+{
+    struct Team {
+        address leader;
+        string name;
+        bytes32 referralCode;
+        address[] members;
+    }
+
+    struct Competition {
+        uint start;
+        uint end;
+        uint maxTeamSize;
+        mapping(address => Team) teams;
+        mapping(string => bool) teamNames;
+        mapping(address => address) memberTeams;
+        mapping(address => address) joinRequests;
+    }
+
+    uint public nextCompetitionIndex = 0;
+    mapping(uint => Competition) public competitions;
+    IReferralStorage public referralStorage;
+
+    event TeamCreated(uint index, address leader, string name, bytes32 referral);
+    event JoinRequestCreated(uint index, address member, address leader);
+    event JoinRequestCanceled(uint index, address member);
+    event JoinRequestApproved(uint index, address member, address leader);
+    event MemberRemoved(uint index, address leader, address member);
+    event CompetitionCreated(uint index, uint start, uint end, uint maxTeamSize);
+    event CompetitionUpdated(uint index, uint start, uint end, uint maxTeamSize);
```

Comment:
i think it might be useful to also remove competition (in case time < start time)